drivers/axi_core/spi_engine/spi_engine.h

```diff
@@ -0,0 +1,188 @@
+ */
+typedef struct spi_engine_desc {
+	/** Type of implementation */
+	enum xil_spi_type	type;
+	/** Pointer to a DMAC used in transmission */
+	struct axi_dmac		*offload_tx_dma;
+	/** Pointer to a DMAC used in reception */
+	struct axi_dmac		*offload_rx_dma;
+	/** Offload's module transfer direction : TX, RX or both */
+	uint8_t			offload_config;
+	/** Number of words that the module has to send */
+	uint8_t			offload_tx_len;
+	/** Number of words that the module has to receive */
+	uint8_t			offload_rx_len;
+	/** Base address where the HDL core is situated */
+	uint32_t		spi_engine_baseaddr;
+	/** Base address where the RX DMAC core is situated */
+	uint32_t		rx_dma_baseaddr;
+	/** Base address where the TX DMAC core is situated */
+	uint32_t		tx_dma_baseaddr;
+	/** Delay between the CS toggle and the start of SCLK */
+	uint8_t			cs_delay;
+	/** Clock divider used in transmission delays */
+	uint32_t		clk_div;
+	/** Data with of one SPI transfer ( in bits ) */
+	uint8_t			data_width;
+	/** The maximum data width supported by the engine */
+	uint8_t 		max_data_width;
+} spi_engine_desc;
+
+
+/**
+ * @struct spi_engine_offload_init_param
+ * @brief  Structure containing the init parameters needed by the offload module
+ */
+typedef struct spi_engine_offload_init_param {
+	/** Base address where the RX DMAC core is situated */
+	uint32_t	rx_dma_baseaddr;
+	/** Base address where the TX DMAC core is situated */
+	uint32_t	tx_dma_baseaddr;
+	/** Offload's module transfer direction : TX, RX or both */
+	uint8_t		offload_config;
+} spi_engine_offload_init_param;
+
+/**
+ * @struct spi_engine_offload_message
+ * @brief  Structure representing an offload message
+ */
+typedef struct spi_engine_offload_message {
+	/** Pointer to the SPI engine commands buffer that will be sent */
+	uint32_t *commands;
+	/** Length of the SPI engine commands buffer */
+	uint32_t no_commands;
+	/** Pointer of the data that will be sent over spi */
+	uint8_t *commands_data;
+	/** The address where the data that will be transmitted is situated */
+	uint32_t tx_addr;
+	/** The address where the data that will be received is situated */
+	uint32_t rx_addr;
+} spi_engine_offload_message;
+
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+
+/* Initialize the SPI engine device */
+int32_t spi_engine_init(struct spi_desc **desc,
+			const struct spi_init_param *param);
+
+/* Write and read data over SPI using the SPI engine */
+int32_t spi_engine_write_and_read(spi_desc *desc,
+				  uint8_t *data,
+				  uint8_t bytes_number);
+
+/* Free the resources used by the SPI engine device */
+int32_t spi_engine_remove(spi_desc *desc);
+
+/* Initialize the SPI engine offload module */
+int32_t spi_engine_offload_init(struct spi_desc *desc,
```

Comment:
Could this be inside spi_engine_init and pass `spi_engine_offload_init_param` as a field of spi_engine_init_param?
In the case the field is null make no offload init.
I'am asking this because there is no spi_engine_offload_remove and the remove for this is done in spi_engine_remove.
Also it seems the spi_engine_offload_transfer uses the spi_desc not an descriptor from offload_init. It looks a little puzzling to me.

Reply:
There are two structures (spi_engine and spi_engine_offload) because there are two independent modes in which the engine can work.